src/goto-instrument/code_contracts.cpp

```diff
@@ -188,6 +205,37 @@ void code_contractst::check_apply_invariants(
     converter.goto_convert(assumption, havoc_code, mode);
   }
 
+  // Temproary variables that store the variant before and after the loop
+  symbol_exprt old_variant_symbol =
+    new_tmp_symbol(
+      variant.type(), loop_head->source_location, function_name, mode)
+      .symbol_expr();
+  symbol_exprt new_variant_symbol =
+    new_tmp_symbol(
+      variant.type(), loop_head->source_location, function_name, mode)
+      .symbol_expr();
+
+  if(!variant.is_nil())
+  {
+    // Generate: a declaration of the temporary variable that stores the loop
+    // variant before the loop entry
+    havoc_code.add(
+      goto_programt::make_decl(old_variant_symbol, loop_head->source_location));
+
+    // Generate: an assignment to store the loop variant before the loop entry
+    // We would like to set a comment associated with a source location.
+    // However, we have already used loop_head->source_location to set a
+    // comment. Furthermore, the third argument of make_assignment is passed by
+    // reference rather than by value. Therefore, it is necessary to clone
+    // loop_head->source_location.
+    source_locationt declaration_old_variant_source_location{
+      loop_head->source_location};
+    havoc_code.add(goto_programt::make_assignment(
+      old_variant_symbol, variant, declaration_old_variant_source_location));
```

Comment:
Instead of generating a goto instruction directly, we should be calling `goto_convert` to properly translate a high-level assignment to appropriate goto instructions. Use:

```c++
    code_assignt old_variant_assignment{old_variant_symbol, variant};
    old_variant_assignment.add_source_location() = loop_head->source_location;
    converter.goto_convert(old_variant_assignment, havoc_code, mode);
```

(same for the other assignment)